### PR TITLE
Use WeakRef to store records on transactions.

### LINKED ISSUE #15549: Unbounded memory growth when creating lots of AR objects inside transaction
Tested on 4.0.5, 4.1.1, and master. On Ubuntu 12.04 32bit, PostgreSQL 9.3, and Ruby 2.1.1 (github edition).

*The interesting thing is, if the AR objects aren't created inside a transaction, there's not as much memory growth.*

```ruby
require 'active_record'
require 'pg'

ActiveRecord::Base.establish_connection("postgres://localhost/test")

ActiveRecord::Base.connection.execute <<-SQL
drop table if exists coupons cascade;
drop table if exists sites cascade;
drop table if exists products cascade;
drop table if exists coupons_sites cascade;
drop table if exists coupons_products cascade;
create table coupons (id serial primary key, name text);
create table sites (id serial primary key);
create table products (id serial primary key);
create table coupons_sites (site_id int references sites, coupon_id int references coupons, primary key (site_id, coupon_id));
create table coupons_products (product_id int references products, coupon_id int references coupons, primary key (product_id, coupon_id));
SQL

class Coupon < ActiveRecord::Base
  has_and_belongs_to_many :products
  has_and_belongs_to_many :sites
end

class Product < ActiveRecord::Base
  has_and_belongs_to_many :coupons
end

class Site < ActiveRecord::Base
  has_and_belongs_to_many :coupons
end

Coupon.transaction do
  1.upto(250_000) do |i|
    Coupon.create!
    puts "Created #{ i } coupons..." if i % 500 == 0
  end
end
```

### PR DESCRIPTION
**Pull request by arthurnn**
In order to restore state on records we need to store all records
touched in the transaction. However if we store all records, we will be
holding a hard reference to them, not allowing them to be garbage
collected. #9068 kinda solved the GC issue inversing the dependency.
As we are on ruby 2.2+ we can use WeakRef, and wont need to inverse that
depedency to restore state.

[fixes #15549] - Because records with a 'create' callback were still
been stored, and memory grow was still a problem for those.

review @chancancode @jeremy @rafaelfranca @sgrif @wangjohn 

All tests are green, and you can run this https://github.com/rails/rails/issues/15549#issuecomment-46035848 , and memory will not grow like crazy.

### CODE REVIEW

**Comment by chancancode** [comment]
:heart: !

1. Do we know how the other Ruby implementations are doing wrt `WeakRef` implementation? @robin850?

2. Have you checked the peak memory usage for before/after without weakref/after with weakref to be sure? (Just because you can't tell from the tests passing.) This might be tricky to tell because MRI doesn't actually release the memory back to the OS. So I think you might need to run a transaction with a few thousand AR objects, then check the number of live objects in the middle or something like that...

3. There is a long standing mystery of why `has_transactional_callbacks?` need to check for `_create_callbacks` too, do you happen to have a clue after working on this? (I think tests actually fails if you remove it, but I haven't fully understood why yet)

cc @tenderlove too

File: activerecord/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,11 @@
+*   Use WeakRef to store records on transactions in order to restore state.
+    Also this avoids a massive memory grow, due the nature of WeakRef, that
+    would allow re records to be garbage collected.
```
Comment (vipulnsward) [inline activerecord/CHANGELOG.md]: extra `re` ?

File: activerecord/lib/active_record/connection_adapters/abstract/transaction.rb [hunk 2]
```diff
@@ -44,21 +45,23 @@ def add_record(record); end
...
 
-      attr_reader :connection, :state, :records, :savepoint_name
+      attr_reader :connection, :state, :savepoint_name, :transactional_records
       attr_writer :joinable
 
       def initialize(connection, options)
         @connection = connection
         @state = TransactionState.new
-        @records = []
+        @transactional_records, @records = [], []
```
Comment (vipulnsward) [inline activerecord/lib/active_record/connection_adapters/abstract/transaction.rb]: `@records` looks unused.

File: activerecord/CHANGELOG.md (hunk 1 above)
Comment (arthurnn) [inline activerecord/CHANGELOG.md]: thanks for the catch.. I pushed a fix.

**Comment by arthurnn** [comment]
2. Yes, I checked using your script from the issue. Memory dont grow a lot as expected, it only grows a little bit as the WeakRef instance itself is not garbage collected, but that is a small object.

3. `_create_callbacks` was on that list because the work-around(pushing the txn state to the records) didnt work for it. Not sure why tho.

**Comment by brainopia** [comment]
Nice one :+1:

File: activerecord/lib/active_record/connection_adapters/abstract/transaction.rb (hunk 2 above)
Comment (bogdan) [inline activerecord/lib/active_record/connection_adapters/abstract/transaction.rb]: Maybe better to use `Set` instead of `Array` to track uniqueness automatically. It may perform faster.

Comment (arthurnn) [inline activerecord/lib/active_record/connection_adapters/abstract/transaction.rb]: good thought, but I would say `Set` would perform worse in this case, as the `push` method is slower on a Set and these list a pretty hot on insert.
Anyways, I think it is worth benchmarking and see some numbers, after I merge this, I will benchmark and see which one is better.thanks

**Comment by tenderlove** [comment]
Is there some reason we don't [use @willbryant's commits](https://github.com/rails/rails/issues/15549#issuecomment-56117367)? I'd rather avoid pulling in weakref if possible.

**Comment by arthurnn** [comment]
@tenderlove we could merge his patch indeed, but the code i am removing here is pretty buggy in general, and has caused lots of problem by it self. I remember seeing some commits from @sgrif to fix problems related to freezed records on this code path.
The weak-ref is exactly what we want in here. And using that make the implementation super simpler, that why I think we should use it.

File: activerecord/lib/active_record/connection_adapters/abstract/transaction.rb [hunk 3]
```diff
@@ -90,6 +98,11 @@ def commit_records
         ite.each do |i|
           i.committed!(should_run_callbacks: false)
         end
+        @records.uniq.each do |i|
+          if i.weakref_alive?
+            i.committed!(should_run_callbacks: false)
+          end
```
Comment (tenderlove) [inline activerecord/lib/active_record/connection_adapters/abstract/transaction.rb]: Does this mean that only objects which *haven't* been GC'd will get the `committed!` callback?

**Comment by jeremy** [comment]
Considering that https://github.com/powershop/rails/commit/2e98a307fa2d7f9a177f4234a9839b686d56a041 fixes #15549, let's evaluate this as a separate design change rather than as a bug fix.

File: activerecord/lib/active_record/connection_adapters/abstract/transaction.rb (hunk 3 above)
Comment (arthurnn) [inline activerecord/lib/active_record/connection_adapters/abstract/transaction.rb]: `committed!` method do 2 things. Restore state on the record, and call de callback. In here I am only restoring the state as I am passing `should_run_callbacks: false`. 
That said. Yes only non-gc records will receive that call, as those are the one we care, as someone else is holding a reference to them, so it needs to have the latest state.

**Comment by sgrif** [comment]
The code I've touched was only tangentially related so I wouldn't take my commits as evidence of anything being needed or not.

**Comment by arthurnn** [comment]
@jeremy yes, I am ok with merge that fix first and this design change afterwards. But this simplify code a lot so I think it is worth it!

**Comment by robin850** [comment]
Hi there!

> Do we know how the other Ruby implementations are doing wrt WeakRef implementation?

Well, at least both `test/cases/transaction_callbacks_test.rb` and `test/cases/transaction_isolation_test.rb` pass on Rubinius (2.4.1) with the postgresql, mysql, mysql2 and sqlite3 adapters. Running the tests on JRuby is a bit more complicated because the Active Record JDBC adapters aren't yet compatible with the stuff introduced in Rails 4.2, but at least the first aforementioned test passes on sqlite3 with JRuby 9000.pre1.

Honestly I'm not comfortable with Active Record related stuff so I don't know if there are other tests to run (or if I ran the good ones).

Regarding memory, here are graphs that show memory evolution when the number of records grow. These results are based on [Godfrey's script](https://github.com/rails/rails/issues/15549#issuecomment-46035848) ; I just took 75e3 records instead of 250e3. The first one is against current `master` and the second, against Arthur's branch:

normal

weakref

There's clearly a reduction of the memory usage with this patch for all implementations ! By the way, I guess it's worth mentioning that I had to pass the `-J-Xmx` option to manually increase the allocated memory running against JRuby on master but I hadn't to with this branch.

Thanks for taking into consideration the support of other implementations guys! :heart:

File: activerecord/lib/active_record/connection_adapters/abstract/transaction.rb (hunk 2 above)
Comment (sgrif) [inline activerecord/lib/active_record/connection_adapters/abstract/transaction.rb]: Yeah, a single `uniq` call should have, at worst, identical performance to a set. If we were doing multiple `uniq` or `include?` calls it would be worth switching to a set. Array will likely remain the best performer here.

File: activerecord/lib/active_record/connection_adapters/abstract/transaction.rb (hunk 3 above)
Comment (sgrif) [inline activerecord/lib/active_record/connection_adapters/abstract/transaction.rb]: What state is getting restored? Should we look at why something needs to happen here?

File: activerecord/lib/active_record/connection_adapters/abstract/transaction.rb [hunk 4]
```diff
@@ -75,6 +78,11 @@ def rollback_records
         ite.each do |i|
           i.rolledback!(force_restore_state: full_rollback?, should_run_callbacks: false)
         end
+        @records.uniq.each do |r|
+          if r.weakref_alive?
```
Comment (sgrif) [inline activerecord/lib/active_record/connection_adapters/abstract/transaction.rb]: What do you think about pulling this out into a method?

```ruby
def each_record
  @records.uniq.each do |record|
    if record.weakref_alive?
      yield record
    end
  end
end
```

**Comment by sgrif** [comment]
While I do think it is worth re-evaluating the original design of this code, I think this is definitely better code than what we had before. It simplifies what we have today, and improves performance. I think we should make this change, and then re-evaluate the design separately.

File: activerecord/lib/active_record/connection_adapters/abstract/transaction.rb (hunk 4 above)
Comment (arthurnn) [inline activerecord/lib/active_record/connection_adapters/abstract/transaction.rb]: this is a good idea.

File: activerecord/lib/active_record/connection_adapters/abstract/transaction.rb (hunk 3 above)
Comment (arthurnn) [inline activerecord/lib/active_record/connection_adapters/abstract/transaction.rb]: this is the state it restores https://github.com/rails/rails/blob/master/activerecord/lib/active_record/transactions.rb#L306

**Comment by tenderlove** [comment]
@robin850 can you do the same test, but with code like this:

```ruby
User.transaction do
  loop do
    User.create!
  end
end
```

Then test against master and this PR?

**Comment by robin850** [comment]
@tenderlove : Sorry I'm not sure to understand ; do you mean with a different loop ? If yes, here are the [master graphs](http://robin.alwaysdata.net/rails/master.html) and here the [graphs using WeakRef](http://robin.alwaysdata.net/rails/weakref.html). :-)

**Comment by tenderlove** [comment]
Sorry, I mean do this graph, but with an AR object that doesn't have any associations, and compare master with this PR.

**Comment by robin850** [comment]
@tenderlove : No problem! Indeed, the results are somewhat less interesting for objects [without any associations](http://robin.alwaysdata.net/rails/single.html) (especially for MRI) than [with relations](http://robin.alwaysdata.net/rails/has_many.html).

**Comment by tenderlove** [comment]
@robin850 note that this PR shows linear memory growth with WeakRefs on MRI vs the flat line on master.  IOW, this PR introduces a memory leak for "less interesting" objects.

**Comment by arthurnn** [comment]
@tenderlove look at the scales. 
This PR, on `"less interesting" objects.` only grows to 17MB under 70k objects.

Without this PR, in objects with relations it grows to 280MB for 70k.

Yes, this PR add a small overhead, but thats super small!

**Comment by tenderlove** [comment]
@arthurnn ya, it is super small, but it's still a leak where we had no leak before.  I've merged powershop@2e98a30, so #15549 should be fixed on master (@robin850 can you make these pretty graphs again?)  We should consider this PR a design change rather than bug fix.

I spelled out in CF exactly why I'm opposed to using WeakRefs.  I'm willing to change my mind, but I need a more convincing argument.  I'd rather have slightly complex code that is deterministic, than slightly more simple code that is non-deterministic.  The risk of "non-deterministic code" outweighs the simplification to me.

**Comment by arthurnn** [comment]
@tenderlove I agree, this is NOT a bug fix PR, it is a design change one. Thats how we should look at it.

What we are holding weak-refs for, is to restore a state, only that. That restore code, has no callback calls to the actual model, or any extra call, it is a simple internal-hash-clear-values operation. https://github.com/rails/rails/blob/master/activerecord/lib/active_record/transactions.rb#L371-L373
That said, GC is non-deterministic, but we guarantee that we will run that restore for the objects that are still alive, in memory, after the transaction is over. Which is exactly the propose of that restore code.
The current code it is really hard to follow, because we inverse the dependencies. Instead of the transaction holding the records, we have a reference to the transaction inside the record. And we need to make sure we call `sync_with_transaction_state` on every accessor, so it can restore its state.

> it is super small, but it's still a leak where we had no leak before

It is not an actual leak. After the transaction is over(committed or rollbacked), those records will not be there anymore.

**Comment by tenderlove** [comment]
> It is not an actual leak. After the transaction is over(committed or rollbacked), those records will not be there anymore.

It is a leak the same way as when we were leaking AR objects.  It's just that WeakRef objects are smaller than AR objects so the leak is more slow.  Technically we were not "leaking" AR objects either.

**Comment by tenderlove** [comment]
We should also test the performance of doing a large txn.  I suspect this PR will slow it down since it has to iterate over every weakref after the txn finishes.  :-/

**Comment by jeremy** [comment]
Good point re. performance characteristics. We're switching from lazy-syncing records with their txn state to eagerly syncing them — wasted effort if we don't commonly work with those records thereafter. What's our breakeven workload, where eager sync is a perf win over lazy, look like? Is it the common case?

**Comment by tenderlove** [comment]
Well, also we can't predict the performance.  If you create 100k objects, and don't reference any of them, the performance will change depending on whether or not GC has run.  This means we could get bug reports that are like "Sometimes it's slow". Randomly executing methods, and random performance characteristics. *shudder*

**Comment by arthurnn** [comment]
I am reopening this because after we merged 045c77c1bc8f9acdc7efe34ba7af59acfe6f955f, it fixed the problem, however we had to add 58410b3d566e6b93c7b71c0eec0fc11ec906b68e, which created the same issue when doing a `Post.transaction { Post.all.take(100_000).destroy }`. Of course the destroy problem is smaller as this is more rarer to happen, but still an issue.
We need to `add_transaction_record` on destroy, because we need to restore the frozen status on those records.

This implementation make easier to cover all those corner cases, as we will store all records and restore state in all. Also code will be simpler. 

The drawback about this, is indeed we are trading memory for CPU, so in end of the transaction we will have to iterate over all records to restore state on them, even if we dont use them.

@jeremy had a good point though:
> What's our breakeven workload, where eager sync is a perf win over lazy, look like? Is it the common case?

I think the common use case is not to touch the records after the transaction, which means the lazy-restore solution is better.

Basically, 
I re-open this, but I will close it again :smile_cat: , as I am assuming the lazy-restore is better due to the common case.
